File: gatebox/wallet/utils/observabilidade/middleware.go

package observabilidade

import (
	"context"
	"time"

	"github.com/labstack/echo/v4"
)

// HTTPMetricsMiddleware middleware básico para coletar métricas HTTP
func HTTPMetricsMiddleware(usecase string) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			start := time.Now()

			// Processar requisição
			err := next(c)

			// Registrar métricas
			duration := time.Since(start)
			method := c.Request().Method
			path := c.Path()
			statusCode := c.Response().Status

			RecordHTTPRequest(
				c.Request().Context(),
				usecase,
				method,
				path,
				duration,
				statusCode,
			)

			return err
		}
	}
}

// EnhancedHTTPMetricsMiddleware middleware avançado para métricas HTTP
func EnhancedHTTPMetricsMiddleware(usecase string) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			// Criar context com timeout se não existir
			ctx := c.Request().Context()
			if _, hasDeadline := ctx.Deadline(); !hasDeadline {
				var cancel context.CancelFunc
				ctx, cancel = context.WithTimeout(ctx, 30*time.Second)
				defer cancel()
				c.SetRequest(c.Request().WithContext(ctx))
			}

			// Timing da camada handler
			handlerTiming := NewLayerTiming(usecase, c.Path(), "handler")

			// Detector de timeout
			timeoutDetector := NewTimeoutDetector(usecase, c.Path())

			start := time.Now()

			// Processar requisição
			err := next(c)

			// Verificar timeout
			timeoutDetector.CheckTimeout(ctx)

			// Finalizar timing do handler
			handlerTiming.Finish(ctx, err)

			// Registrar métricas HTTP originais
			duration := time.Since(start)
			method := c.Request().Method
			path := c.Path()
			statusCode := c.Response().Status

			RecordHTTPRequest(
				ctx,
				usecase,
				method,
				path,
				duration,
				statusCode,
			)

			return err
		}
	}
}

File: gatebox/gatebox-rust/src/observabilidade/middleware.rs
// Converted from gateboxgo/utils/observabilidade/middleware.go (HTTP metrics for Axum)
use axum::{
    extract::Request,
    middleware::Next,
    response::Response,
};
use std::time::Instant;
use tower_http::timeout::TimeoutLayer;

use crate::observabilidade::metrics;

#[allow(dead_code)]
pub fn timeout_layer(timeout: std::time::Duration) -> TimeoutLayer {
    TimeoutLayer::new(timeout)
}

/// Middleware that records HTTP metrics. Use with tower::ServiceBuilder and a captured usecase.
pub async fn enhanced_http_metrics_layer(
    usecase: String,
    request: Request,
    next: Next,
) -> Response {
    let method = request.method().to_string();
    let path = request.uri().path().to_string();
    let start = Instant::now();

    let response = next.run(request).await;
    let status_code = response.status().as_u16();
    let duration = start.elapsed();

    metrics::record_http_request(&usecase, &method, &path, duration, status_code);

    response
}
